ws-server/src/server/module.rs

```diff
@@ -0,0 +1,162 @@
+use super::*;
+
+#[derive(Default)]
+pub struct RpcModule {
+	methods: Methods,
+}
+
+impl RpcModule {
+	/// Instantiate a new `RpcModule`.
+	pub fn new() -> Self {
+		RpcModule { methods: Methods::default() }
+	}
+
+	/// Add context for this module, turning it into an `RpcContextModule`.
+	pub fn with_context<Context>(self, ctx: Context) -> RpcContextModule<Context> {
+		RpcContextModule { ctx: Arc::new(ctx), module: self }
+	}
+
+	fn verify_method_name(&mut self, name: &str) -> Result<(), Error> {
+		if self.methods.get(name).is_some() {
+			return Err(Error::MethodAlreadyRegistered(name.into()));
+		}
+
+		Ok(())
+	}
+
+	/// Register a new RPC method, which responds with a given callback.
+	pub fn register_method<F, R>(&mut self, method_name: &'static str, callback: F) -> Result<(), Error>
+	where
+		R: Serialize,
+		F: RpcMethod<R>,
+	{
+		self.verify_method_name(method_name)?;
+
+		self.methods.insert(
+			method_name,
+			Box::new(move |id, params, tx, _| {
+				let result = callback(params)?;
+
+				send_response(id, tx, result);
+
+				Ok(())
+			}),
+		);
+
+		Ok(())
+	}
+
+	/// Register a new RPC subscription, with subscribe and unsubscribe methods.
+	pub fn register_subscription(
+		&mut self,
+		subscribe_method_name: &'static str,
+		unsubscribe_method_name: &'static str,
+	) -> Result<SubscriptionSink, Error> {
+		if subscribe_method_name == unsubscribe_method_name {
+			return Err(Error::MethodAlreadyRegistered(subscribe_method_name.into()));
```

Comment:
This should be some other error type I think. The problem isn't really that the method is registered but that the names of the (un-)subscribe methods are the same, `Error::InvalidSubscriptionName` perhaps?